frontend/src/liboperator/operator/operator.go

```diff
@@ -480,6 +484,111 @@ func openwrt(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
 	proxy.ServeHTTP(w, r)
 }
 
+// WebSocket endpoint that proxies ADB
+func adbProxy(w http.ResponseWriter, r *http.Request, pool *DevicePool) {
+	vars := mux.Vars(r)
+	devId := vars["deviceId"]
+	dev := pool.GetDevice(devId)
+
+	if dev == nil {
+		http.Error(w, "Device not found", http.StatusNotFound)
+		return
+	}
+
+	devInfo := dev.privateData.(map[string]interface{})
+
+	// Find adb port for the device.
+	//   1. If registered device info has adb_port, use it.
+	//   2. If not, get index of the device and 6520 + index would be the adb port.
+	adbPort := 0
+	if adb_port, ok := devInfo["adb_port"]; ok {
+		adbPort = int(adb_port.(float64))
+	} else {
+		for i, id := range pool.DeviceIds() {
+			if id == devId {
+				adbPort = 6520 + i
+				break
+			}
+		}
```

Comment:
I think this fallback mechanism is not reliable. Especially when the base instance number is not 1. Shouldn't we make backporting changes for previous CF releases instead?

Reply:
@ser-io @jemoreira I have some question about this.

Does it okay to move this logic to orchestrator? or it needs to be in operator. Since operator does not have adb connection info for the device I added adb port info to device registration info, but it might be a problem when device is launched with old host packages.

I added the fallback logic, but there are some situations this might not work.

But if this logic is in the orchestrator, it might be possible to get adb_serial info from orchestrator.

Do you have any suggestions for that?

Reply:
AFAIK, we would unite Operator and Host Orchestrator into one. (Correct me if I'm wrong.)

Then, it's better to be in Host Orchestrator, since it holds information about adb_serial.
TBH, we don't need ADB port but ADB serial instead, isn't it?

Reply:
If an instance is started by `cvd start` it would be fine. adb_serial for host orchestrator is got by `cvd fleet` which shows instances created by `cvd start` (of course created by host orchestrator).

If supporting ADB connection proxy from host orchestrator it might need to rename `/devices/{deviceId}/adb` endpoint to `cvds/{group}/{name}/adb`.

Or another option might be just return an error `Device is not supported for ADB` when deviceInfo does not have ADB port for operator, and backport ADB port registration change(http://aosp/3154596) for the device when it is needed.

@ser-io @jemoreira That do you think?

Reply:
Decided not to use fallback logic but just return errors when adb_port is not registered.